include/dlaf/eigensolver/tridiag_solver/mc.h

```diff
@@ -0,0 +1,241 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2022, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+#pragma once
+
+#include <algorithm>
+
+#include <pika/future.hpp>
+#include <pika/unwrap.hpp>
+
+#include "dlaf/common/callable_object.h"
+#include "dlaf/eigensolver/tridiag_solver/api.h"
+#include "dlaf/eigensolver/tridiag_solver/merge.h"
+#include "dlaf/lapack/tile.h"
+#include "dlaf/matrix/copy_tile.h"
+#include "dlaf/sender/make_sender_algorithm_overloads.h"
+#include "dlaf/sender/policy.h"
+#include "dlaf/sender/transform.h"
+#include "dlaf/types.h"
+
+namespace dlaf::eigensolver::internal {
+
+template <class T>
+struct TridiagSolver<Backend::MC, Device::CPU, T> {
+  static void call(Matrix<T, Device::CPU>& mat_trd, Matrix<T, Device::CPU>& d,
+                   Matrix<T, Device::CPU>& mat_ev);
+};
+
+/// Splits [i_begin, i_end] in the middle and waits for all splits on [i_begin, i_middle] and [i_middle +
+/// 1, i_end] before saving the triad <i_begin, i_middle, i_end> into `indices`.
+///
+/// The recursive calls span a binary tree which is traversed in depth first left-right-root order. That
+/// is also the order of triads in `indices`.
+///
+/// Note: the intervals are all closed!
+///
+inline void splitIntervalInTheMiddleRecursively(
+    SizeType i_begin, SizeType i_end, std::vector<std::tuple<SizeType, SizeType, SizeType>>& indices) {
```

Comment:
I have the same comment about `std::tuple<SizeType, SizeType, SizeType>` as about the "sync matrix", i.e. it'd be really nice to give names to those `SizeType`s by using a struct instead.

Reply:
Sounds good and fine by me to do that in a separate PR.

Reply:
Just another suggestion along these lines for refactoring activity.

It may be also nice to define a type for closed intervals. Currently they are represented as two independent `SizeType`, but having them merged in a struct (possibly with some checks, e.g. `i_begin <= i_end`, and some functions, like `split` perhaps) may improve API (currently it is a lot "C-style") and maintainability.

I don't know yet if they can be extended to be useful also in-place of `std::tuple<SizeType, SizeType, SizeType>` which seems to represent a closed range with a pivot point, but that can be reasoned about too.

Reply:
I suggest we open an issue about this and schedule the cleanup for another PR, so as not to block this further.